Add Learn page on zero-knowledge proofs (#6890)
ethereum/ethereum-org-website: src/content/zero-knowledge-proofs.md

Credit card payments are often visible to multiple parties, including the payments provider, banks, and other interested parties (e.g., government authorities). While financial surveillance has benefits for identifying illegal activity, it also undermines the privacy of ordinary citizens.

Cryptocurrencies were intended to provide a means for users to conduct private, peer-to-peer transactions. But most cryptocurrency transactions are openly visible on public blockchains. User identities are often pseudonymous and either wilfully linked to real-world identities (e.g. by including ETH addresses on Twitter or Github profiles) or can be associated with real-world identities using basic on and off-chain data analysis. 

There are specific “privacy coins” designed for completely anonymous transactions. Privacy-focused blockchains, such as ZCash and Monero, shield transaction details, including sender/receiver addresses, asset type, quantity, and the transaction timeline.

By baking in zero-knowledge technology into the protocol, privacy-focused blockchain networks allow nodes to validate transactions without needing to access transaction data. 

> Zero-knowledge proofs are also being applied to anonymizing transactions on public blockchains. An example is Tornado Cash, a decentralized, non-custodial service that allows users to conduct private transactions on Ethereum. Tornado Cash uses zero-knowledge proofs to obfuscate transaction details and guarantee financial privacy.  Unfortunately, because these are "opt-in" privacy tools they are associated with illicit activity. To overcome this, privacy has to eventually become the default on public blockchains.

Reviewer: > Unfortunately, because these are "opt-in" privacy tools they are associated with illicit activity. To overcome this, privacy has to eventually become the default on public blockchains.

It may be worth mentioning a note here about plausible deniability. By placing privacy at the application level, the simple fact that you're using that application  can raise unwanted attention, and it's difficult to deny that you're using it for that reason or that the privacy matters when using something like Torando cash, which has the sole purpose of making funds difficult to track. As you state, placing privacy as the default overcomes this.

Author: It's acc a good point.  For example, Tornado Cash is usually in the news each time some hacker is trying to launder stolen funds (e.g., what happened with the Harmony One token bridge hack). That said, I think @jmcook1186's description may be enough--talking  about plausible deniability may be too much info imo.

---
title: Zero-knowledge proofs 
description: An non-technical introduction to zero-knowledge proofs for beginners. 
lang: en
sidebar: true
---

## What are zero-knowledge proofs?

A zero-knowledge proof is a way of proving the validity of a statement without revealing the statement itself. The ‘prover’ is the party trying to prove a claim, while the ‘verifier’ is responsible for validating the claim. 

> Zero-knowledge proofs first appeared in a 1985 paper, “The knowledge complexity of interactive proof systems”, co-authored by MIT researchers Shafi Goldwasser, Silvio Micali and Charles Rackoff. The [paper](http://people.csail.mit.edu/silvio/Selected%20Scientific%20Papers/Proof%20Systems/The_Knowledge_Complexity_Of_Interactive_Proof_Systems.pdf)] provides a definition of zero-knowledge proofs widely used today:

Reviewer 2: ```suggestion
Zero-knowledge proofs first appeared in a 1985 paper, “[The knowledge complexity of interactive proof systems](http://people.csail.mit.edu/silvio/Selected%20Scientific%20Papers/Proof%20Systems/The_Knowledge_Complexity_Of_Interactive_Proof_Systems.pdf)” which provides a definition of zero-knowledge proofs widely used today:
```

> A zero-knowledge protocol is a method by which one party (the prover) can prove to another party (the verifier) that something is true, without revealing any information apart from the fact that this specific statement is true.

> Zero-knowledge proofs have undergone many improvements over the years. Many of these advances have enhanced the usability and efficiency of zero-knowledge proofs, leading to their applications in various areas. 

Reviewer 2: ```suggestion
Zero-knowledge proofs have improved over the years and they are now being used in several real-world applications. 
```

3. **Zero-knowledge**: The verifier learns nothing about a statement beyond its validity or falsity (they have “zero knowledge” of the statement). This requirement also prevents the verifier from deriving the original input (the statement’s contents) from the proof. 

In basic form, a zero-knowledge proof is made up of three elements: **witness**, **challenge**, and **response**. 

- **Witness**: The witness is the statement to be proven. The proving process starts with creating a set of questions which can only be answered by a party with access to the witness. The prover randomly chooses a question from the set, computes a proof, and sends it to the verifier. 

- **Challenge**: The verifier picks another question from the set and asks the prover to answer it. 

- **Response**: The prover accepts the question, calculates the answer, and returns it to the verifier. The prover’s response allows the verifier to check if the former really has access to the witness. 

> To ensure the prover isn’t guessing blindly, the verifier picks more questions to ask. By repeating this interaction many times, the possibility of the verifier faking knowledge of the witness drops significantly until the verifier is satisfied.

Author: I just added line 48 to line 46 to clear up confusion. It's a continuation of the "Response" section and explains how a verifier repeats the questioning process until the possibility that the prover is lucky drops to an insignificant value. 

Line 50-52 is to provide an on-ramp to the section on non-interactive proofs.

Non-interactive proving reduces communication between prover and verifier, making zk-proofs more efficient. Moreover, once a proof is generated, it is available for anyone else (with access to the shared key and verification algorithm) to verify.

Non-interactive proofs represented a breakthrough for zero-knowledge technology and spurred the development of proving systems used today. We discuss these proof types below:

### Types of zero-knowledge proofs {#types-of-zero-knowledge-proofs}

#### ZK-SNARKs {#zk-snarks}

ZK-SNARK is an acronym for **Zero-Knowledge Succinct Non-Interactive Argument of Knowledge**. The ZK-SNARK protocol has the following qualities:

- **Zero-knowledge**: A verifier can validate the integrity of a statement without knowing anything else about the statement. The only knowledge the verifier has of the statement is whether it is true or false. 

- **Succinct**: The zero-knowledge proof is smaller than the witness and can be verified quickly. 

- **Non-interactive**: The proof is ‘non-interactive’ because the prover and verifier only interact once, unlike interactive proofs that require multiple rounds of communication. 

> - **Argument**: The proof satisfies the ‘soundness’ requirement, so the possibility of a cheating prover succeeding is extremely low. However, this assumes that provers leverage bounded computation. 

Reviewer 2: ```suggestion
- **Argument**: The proof satisfies the ‘soundness’ requirement, so cheating is extremely unlikely. 
```

- **(Of) Knowledge**: The zero-knowledge proof cannot be constructed without access to the secret information (witness). It is difficult, if not impossible, for a prover who doesn’t have the witness to compute a valid zero-knowledge proof. 

The ‘shared key’ mentioned earlier refers to public parameters that the prover and verifier agree to use in generating and verifying proofs. Generating the public parameters (collectively known as the Common Reference String) is a sensitive operation because of its importance in the protocol’s security. If the entropy (randomness) used in generating the CRS gets into the hands of a dishonest prover, they can compute false proofs. 

[Multi-party computation (MPC)](https://en.wikipedia.org/wiki/Secure_multi-party_computation) is a way of reducing the risks in generating public parameters. Multiple parties participate in a [trusted setup ceremony](https://zkproof.org/2021/06/30/setup-ceremonies/amp/), where each person contributes some random values to generate the CRS. As long as one honest party destroys their portion of the entropy, the ZK-SNARK protocol retains computational soundness. 

> Trusted setups require users to trust the integrity of everyone involved in the parameter-generation process, negating the ‘trustlessness’ principle underpinning blockchain designs. However, advanced research has led to the creation of ZK-STARKs, proving protocols that work with a non-trusted setup. 

Reviewer 2: ```suggestion
Trusted setups require users to trust the participants in parameter-generation. However, the development of ZK-STARKs has enabled proving protocols that work with a non-trusted setup. 
```

#### ZK-STARKs {#zk-starks}

> ZK-STARK is an acronym for **Zero-Knowledge Scalable Transparent Argument of Knowledge**. If you notice, a ZK-STARK is similar to a ZK-SNARK, except for two qualities:

Reviewer 2: ```suggestion
ZK-STARK is an acronym for **Zero-Knowledge Scalable Transparent Argument of Knowledge**. ZK-STARKs are similar to ZK-SNARKs, except for that they are:
```

- **Scalable**: ZK-STARK is faster than ZK-SNARK at generating and verifying proofs when the size of the witness is larger. With STARK proofs, prover and verification times only slightly increase as the witness grows (SNARK prover and verifier times increase linearly with witness size). 

- **Transparent**: ZK-STARK relies on publicly verifiable randomness to generate public parameters for proving and verification instead of a trusted setup. Thus, they are more transparent compared to ZK-SNARKs. 

> While ZK-STARK improves on ZK-SNARK, it has a significant drawback: it produces larger proofs. As a result, STARK proofs have higher verification overhead—although this arguably becomes less of an issue when generating proofs for large datasets. In this case, a ZK-STARK may be more cost-effective than a ZK-SNARK. 

Reviewer 2: ```suggestion
ZK-STARKs produce larger proofs than ZK-SNARKs meaning they generally have higher verification overheads. However, there are cases (such as proving large datasets) where ZK-STARKs may be more cost-effective than ZK-SNARKs. 
```

## Use-cases for zero-knowledge proofs {#use-cases-for-zero-knowledge-proofs}

### Anonymous payments {#anonymous-payments} 

Credit card payments are often visible to multiple parties, including the payments provider, your bank, and other interested parties (e.g., government authorities). While financial surveillance is good for curbing illegal financial activities, it poses risks for ordinary citizens. Such individuals may wish to keep details of their transactions away from prying eyes. 

> Cryptocurrencies were designed to provide a means for users to conduct private, peer-to-peer transactions. But cryptos only offer marginal improvements on financial privacy, as transaction details are publicly visible on the blockchain. Also, user identities are pseudonymous, not completely anonymous, allowing cutting-edge blockchain analysis tools to correlate wallet addresses with real-life identities. 

Reviewer 2: ```suggestion
Cryptocurrencies were intended to provide a means for users to conduct private, peer-to-peer transactions. But most cryptocurrency transactions are openly visible on public blockchains. User identities are often pseudonymous and either wilfully linked to real-world identities (e.g. by including ETH addresses on Twitter or Github profiles) or can be associated with real-world identities using basic on and off-chain data analysis. 
```

> However, continued research into financial privacy techniques led to the development of “privacy coins”, which are cryptocurrencies specifically designed for completely anonymous transactions. Privacy-focused blockchains, such as ZCash and Monero, shield transaction details, including sender/receiver addresses, asset type, quantity, and the transaction timeline. 

Reviewer 2: Changing this because I don't think it is research chronology that explains the privacy/non-privacy coin dynamic. ZCash, Monero etc have been around for ages and it is not newness that holds them back. IMO, it is lack of demand combined with the illegality connotations of choosing privacy tools.

```suggestion
There are specific “privacy coins” designed for completely anonymous transactions. Privacy-focused blockchains, such as ZCash and Monero, shield transaction details, including sender/receiver addresses, asset type, quantity, and the transaction timeline.
```

Some will ask, “How can blockchain nodes verify the validity of transactions without accessing transaction details?”. This is where zero-knowledge proofs come into the picture. By baking in zero-knowledge technology into the protocol, privacy-focused blockchain networks allow nodes to validate transactions without needing to access transaction data. 

> Zero-knowledge proofs are also being applied to anonymizing transactions on public blockchains. An example is Tornado Cash, a decentralized, non-custodial service that allows users to conduct private transactions on Ethereum. Tornado Cash uses zero-knowledge proofs to obfuscate transaction details and guarantee financial privacy.  

Reviewer 2: ```suggestion
Zero-knowledge proofs are also being applied to anonymizing transactions on public blockchains. An example is Tornado Cash, a decentralized, non-custodial service that allows users to conduct private transactions on Ethereum. Tornado Cash uses zero-knowledge proofs to obfuscate transaction details and guarantee financial privacy.  Unfortunately, because these are "opt-in" privacy tools they are associated with illicit activity. To overcome this, privacy has to eventually become the default on public blockchains.
```

### Identity protection {#identity-protection}

> The scenario described earlier (proving citizenship) is an example of how current identity management systems put personal information at risk. But, as explained subsequently, zero-knowledge proofs can help individuals validate identity whilst protecting sensitive details. 

Reviewer 2: ```suggestion
Current identity management systems put personal information at risk. Zero-knowledge proofs can help individuals validate identity whilst protecting sensitive details. 
```

Verifiable computation is critical to improving processing speeds on blockchains without reducing security. Understanding this requires knowing the differences in proposed solutions for scaling Ethereum. 

[On-chain scaling solutions](/developers/docs/scaling/#on-chain-scaling), such as sharding, require extensive modification of the blockchain’s base layer. However, this approach is highly complex and errors in implementation can undermine Ethereum’s security model. 

[Off-chain scaling solutions](/developers/docs/scaling/#off-chain-scaling) don’t require redesigning the core Ethereum protocol. Instead they rely on an outsourced computation model to improve throughput on Ethereum’s base layer. 

Here’s how that works in practice:

-  Instead of processing every transaction, Ethereum offloads execution to a separate chain. 

- After processing transactions, the other chain returns the results to be applied to Ethereum’s state.

> The benefit here is that the main Ethereum doesn’t have to do any execution and only needs to apply results from outsourced computation to its state. This reduces network congestion and also improves transaction speeds (off-chain protocols optimize for faster execution).

Author: Missed this comment earlier. I'll look through and use terms consistently per your suggestion.

> But how can Ethereum be sure that off-chain transactions were executed correctly? The chain needs a way to validate off-chain transactions without re-executing them, or else the value of off-chain execution is lost. 

Reviewer 2: ```suggestion
The chain needs a way to validate off-chain transactions without re-executing them, or else the value of off-chain execution is lost. 
```
